app/webhub_fuzz.go

```diff
@@ -0,0 +1,293 @@
+// (Also reduce the number of push notification workers to 1 to debug stack traces easily.)
+// 3. go-fuzz-build github.com/mattermost/mattermost-server/v5/app.
+// 4. Generate a corpus dir. It's just a directory with files containing random data
+// for go-fuzz to use as an initial seed. Use the generateInitialCorpus function for that.
+// 5. go-fuzz -bin=app-fuzz.zip -workdir=./workdir
+var mainHelper *testlib.MainHelper
+
+func init() {
+	testing.Init()
+	var options = testlib.HelperOptions{
+		EnableStore:     true,
+		EnableResources: true,
+	}
+
+	mainHelper = testlib.NewMainHelperWithOptions(&options)
+	// defer mainHelper.Close()
+	// mainHelper.Main(m)
+}
+
+func dummyWebsocketHandler() http.HandlerFunc {
+	return func(w http.ResponseWriter, req *http.Request) {
+		upgrader := &websocket.Upgrader{
+			ReadBufferSize:  1024,
+			WriteBufferSize: 1024,
+		}
+		conn, err := upgrader.Upgrade(w, req, nil)
+		for err == nil {
+			_, _, err = conn.ReadMessage()
+		}
+		if _, ok := err.(*websocket.CloseError); !ok {
+			panic(err)
+		}
+	}
+}
+
+func registerDummyWebConn(a *App, addr net.Addr, userId string) *WebConn {
+	session, appErr := a.CreateSession(&model.Session{
+		UserId: userId,
+	})
+	if appErr != nil {
+		panic(appErr)
+	}
+
+	d := websocket.Dialer{}
+	c, _, err := d.Dial("ws://"+addr.String()+"/ws", nil)
+	if err != nil {
+		panic(err)
+	}
+
+	wc := a.NewWebConn(c, *session, goi18n.IdentityTfunc(), "en")
+	a.HubRegister(wc)
+	go wc.Pump()
+	return wc
+}
+
+type actionData struct {
+	event                string
+	createUserID         string
+	selectChannelID      string
+	selectTeamID         string
+	invalidateConnUserID string
+	updateConnUserID     string
+	attachment           map[string]interface{}
+}
+
+func getActionData(data []byte, userIDs, teamIDs, channelIDs []string) *actionData {
+	// Some sample events
+	events := []string{
+		model.WEBSOCKET_EVENT_CHANNEL_CREATED,
+		model.WEBSOCKET_EVENT_CHANNEL_DELETED,
+		model.WEBSOCKET_EVENT_USER_ADDED,
+		model.WEBSOCKET_EVENT_USER_UPDATED,
+		model.WEBSOCKET_EVENT_STATUS_CHANGE,
+		model.WEBSOCKET_EVENT_HELLO,
+		model.WEBSOCKET_AUTHENTICATION_CHALLENGE,
+		model.WEBSOCKET_EVENT_REACTION_ADDED,
+		model.WEBSOCKET_EVENT_REACTION_REMOVED,
+		model.WEBSOCKET_EVENT_RESPONSE,
+	}
+	// We need atleast 10 bytes to get all the data we need
+	if len(data) < 10 {
+		return nil
+	}
+	input := &actionData{}
+	//	Assign userID, channelID, teamID randomly from respective byte indices
+	input.createUserID = userIDs[int(data[0])%len(userIDs)]
+	input.selectChannelID = channelIDs[int(data[1])%len(channelIDs)]
+	input.selectTeamID = teamIDs[int(data[2])%len(teamIDs)]
+	input.invalidateConnUserID = userIDs[int(data[3])%len(userIDs)]
+	input.updateConnUserID = userIDs[int(data[4])%len(userIDs)]
+	input.event = events[int(data[5])%len(events)]
+	data = data[6:]
+	input.attachment = make(map[string]interface{})
+	for len(data) >= 4 { // 2 bytes key, 2 bytes value
```

Comment:
Out of curiosity: Is this the right approach to discard last bytes if we have 3 bytes remaining? I mean can we guarantee that we never get inputs like this?

Reply:
We are just trying to craft an input for our program. No we can't guarantee, and we don't bother much. We will let go-fuzz figure out the next input to give based on our return codes and the coverage analysis that it does.

Reply:
So basically we are treating `[10]byte{...}` and `[12]byte{...}` same.

Reply:
Correct.

Reply:
I see, thanks!